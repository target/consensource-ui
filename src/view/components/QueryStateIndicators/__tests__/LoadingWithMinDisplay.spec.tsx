import React from 'react';
import { render, act, screen } from '@testing-library/react';
import { QueryResult } from 'react-query';
import {
  LoadingWithMinDisplay,
  WAIT_TIME_MS,
  MIN_DISPLAY_TIME_MS,
} from '../LoadingWithMinDisplay';

describe('<LoadingWithMinDisplay />', () => {
  const baseQueryRes = {
    isLoading: true,
    error: undefined,
    data: [],
  } as QueryResult<any>;

  beforeEach(() => {
    jest.useFakeTimers();
  });

  afterEach(() => {
    act(() => {
      jest.runAllTimers();
    });
  });

  const advanceTimersInAct = (advanceTimeMs: number) => {
    act(() => {
      jest.advanceTimersByTime(advanceTimeMs);
    });
  };

  describe('returns null', () => {
    let view: any;

    beforeEach(() => {
      const {
        container: { firstChild },
      } = render(
        <LoadingWithMinDisplay queryRes={baseQueryRes}>
          test
        </LoadingWithMinDisplay>,
      );

      view = firstChild;
    });

    it('on first render', () => {
      expect(view).toBe(null);
    });

    it('while the wait timer is active', () => {
      advanceTimersInAct(WAIT_TIME_MS - 1);
      expect(view).toBe(null);
    });
  });

  describe('returns a loading indicator', () => {
    const loadingTestId = 'loading-indicator';
    const loadingIndicatorElem = <div data-testid={loadingTestId} />;

    beforeEach(() => {
      render(
        <LoadingWithMinDisplay
          queryRes={baseQueryRes}
          loadingIndicator={loadingIndicatorElem}
        >
          test
        </LoadingWithMinDisplay>,
      );
    });

    it('renders a default loading indicator', () => {
      const { container } = render(
        <LoadingWithMinDisplay queryRes={baseQueryRes}>
          test
        </LoadingWithMinDisplay>,
      );

      advanceTimersInAct(WAIT_TIME_MS + 1);
      expect(container).toMatchSnapshot();
    });

    it('if the display timer is active', async () => {
      advanceTimersInAct(WAIT_TIME_MS + 1);
      expect(screen.getByTestId(loadingTestId)).toBeInTheDocument();
    });

    it('waits at least waitTimeMs to display the loading indicator', () => {
      advanceTimersInAct(WAIT_TIME_MS - 1);
      expect(screen.queryByTestId(loadingTestId)).toBe(null);
    });
  });

  describe('returns an error indicator', () => {
    const errorTestId = 'error-indicator';
    const errorIndicatorElem = <div data-testid={errorTestId} />;

    it('renders a default error indicator', () => {
      const queryRes = { ...baseQueryRes, isLoading: false, error: true };

      const { container } = render(
        <LoadingWithMinDisplay queryRes={queryRes}>test</LoadingWithMinDisplay>,
      );

      advanceTimersInAct(WAIT_TIME_MS + 1);
      expect(container).toMatchSnapshot();
    });

    it('if the query res returned an error', async () => {
      const queryRes = { ...baseQueryRes, isLoading: false, error: true };

      render(
        <LoadingWithMinDisplay
          queryRes={queryRes}
          errorIndicator={errorIndicatorElem}
        >
          test
        </LoadingWithMinDisplay>,
      );

      advanceTimersInAct(WAIT_TIME_MS + 1);
      expect(screen.getByTestId(errorTestId)).toBeInTheDocument();
    });

    it('if the query res returned a falsey value', async () => {
      const queryRes = { ...baseQueryRes, isLoading: false, data: undefined };

      render(
        <LoadingWithMinDisplay
          queryRes={queryRes}
          errorIndicator={errorIndicatorElem}
        >
          test
        </LoadingWithMinDisplay>,
      );

      advanceTimersInAct(WAIT_TIME_MS + 1);
      expect(screen.getByTestId(errorTestId)).toBeInTheDocument();
    });
  });

  describe('returns children', () => {
    const childTestId = 'child';
    const childComponent = <div data-testid={childTestId} />;

    it('invokes the children function with query data', async () => {
      const mockFn = jest.fn(() => childComponent);

      render(
        <LoadingWithMinDisplay queryRes={{ ...baseQueryRes, isLoading: false }}>
          {mockFn}
        </LoadingWithMinDisplay>,
      );

      advanceTimersInAct(WAIT_TIME_MS + MIN_DISPLAY_TIME_MS + 1);

      expect(mockFn).toHaveBeenCalledWith(baseQueryRes.data);
      expect(screen.getByTestId(childTestId)).toBeInTheDocument();
    });

    it('returns the children react node', async () => {
      render(
        <LoadingWithMinDisplay queryRes={{ ...baseQueryRes, isLoading: false }}>
          {childComponent}
        </LoadingWithMinDisplay>,
      );

      advanceTimersInAct(WAIT_TIME_MS + MIN_DISPLAY_TIME_MS + 1);
      expect(screen.getByTestId(childTestId)).toBeInTheDocument();
    });

    it('does not render the loading indicator if loading completes before the wait time is over', async () => {
      const { rerender } = render(
        <LoadingWithMinDisplay queryRes={baseQueryRes}>
          {childComponent}
        </LoadingWithMinDisplay>,
      );

      rerender(
        <LoadingWithMinDisplay queryRes={{ ...baseQueryRes, isLoading: false }}>
          {childComponent}
        </LoadingWithMinDisplay>,
      );

      advanceTimersInAct(WAIT_TIME_MS + 1);
      expect(screen.getByTestId(childTestId)).toBeInTheDocument();
    });
  });
});
