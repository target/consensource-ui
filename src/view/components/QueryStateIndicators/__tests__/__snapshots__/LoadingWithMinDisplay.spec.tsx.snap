// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`<LoadingWithMinDisplay /> returns a loading indicator renders a default loading indicator 1`] = `
<div>
  <div
    class="MuiCircularProgress-root MuiCircularProgress-colorPrimary MuiCircularProgress-indeterminate"
    role="progressbar"
    style="width: 40px; height: 40px;"
  >
    <svg
      class="MuiCircularProgress-svg"
      viewBox="22 22 44 44"
    >
      <circle
        class="MuiCircularProgress-circle MuiCircularProgress-circleIndeterminate"
        cx="44"
        cy="44"
        fill="none"
        r="20.2"
        stroke-width="3.6"
      />
    </svg>
  </div>
</div>
`;

exports[`<LoadingWithMinDisplay /> returns an error indicator renders a default error indicator 1`] = `
<div>
  <div
    class="MuiGrid-root MuiGrid-container MuiGrid-spacing-xs-1 MuiGrid-direction-xs-column MuiGrid-align-items-xs-center"
  >
    <div
      class="MuiGrid-root MuiGrid-item"
    >
      <svg
        class="MuiSvgIcon-root MuiSvgIcon-colorError MuiSvgIcon-fontSizeSmall"
        focusable="false"
        role="img"
        viewBox="0 0 24 24"
      >
        <path
          d="M1 21h22L12 2 1 21zm12-3h-2v-2h2v2zm0-4h-2v-4h2v4z"
        />
        <title>
          error
        </title>
      </svg>
    </div>
    <div
      class="MuiGrid-root MuiGrid-item"
    >
      <p
        class="MuiTypography-root MuiTypography-body1"
      />
    </div>
  </div>
</div>
`;
